core-plugins/src/main/java/co/cask/hydrator/plugin/batch/action/HDFSFileMoveAction.java

```diff
@@ -0,0 +1,153 @@
+ * the License.
+ */
+
+package co.cask.hydrator.plugin.batch.action;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.plugin.PluginConfig;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.action.Action;
+import co.cask.cdap.etl.api.action.ActionContext;
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.PathFilter;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+
+/**
+ * Action that moves file(s) within HDFS in the same cluster.
+ * A user must specify file/directory path and destination file/directory path
+ * Optionals include fileRegex
+ */
+@Plugin(type = Action.PLUGIN_TYPE)
+@Name("HDFSFileMoveAction")
+@Description("Action to move files within HDFS")
+public class HDFSFileMoveAction extends Action {
+  private static final Logger LOG = LoggerFactory.getLogger(HDFSFileMoveAction.class);
+
+  private HDFSActionConfig config;
+
+  public HDFSFileMoveAction(HDFSActionConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void run(ActionContext context) throws Exception {
+    Path source = new Path(config.sourcePath);
+    Path dest = new Path(config.destPath);
+    FileSystem fileSystem = source.getFileSystem(new Configuration());
+
+    if (fileSystem.getFileStatus(source).isFile()) { //moving single file
+
+      try {
+        if (!fileSystem.rename(source, dest)) {
+          if (!config.continueOnError) {
+            throw new IOException(String.format("Failed to rename file {} to {}", source.toString(), dest.toString()));
+          }
+          LOG.error("Failed to move file {} to {}", source.toString(), dest.toString());
+        }
+      } catch (IOException e) {
+        if (!config.continueOnError) {
+          throw e;
+        }
+        LOG.error("Failed to move file {} to {}", source.toString(), dest.toString(), e);
+      }
+      return;
+    }
+
+    //moving contents of directory
+    FileStatus[] listFiles;
+    if (config.fileRegex != null) {
+      PathFilter filter = new PathFilter() {
+        private final Pattern pattern = Pattern.compile(config.fileRegex);
+
+        @Override
+        public boolean accept(Path path) {
+          return pattern.matcher(path.getName()).matches();
+        }
+      };
+
+      listFiles = fileSystem.listStatus(source, filter);
+    } else {
+      listFiles = fileSystem.listStatus(source);
+    }
+
+    for (FileStatus file: listFiles) {
+      source = file.getPath();
+      //Moving directory, so destination filePath doesn't have filename
+      dest = new Path(config.destPath);
```

Comment:
creating a new `Path` object will thrown an `IllegalArgumentException` if the path does not exist. I'll add a check to ensure it is a directory though.